### DISCUSSION TITLE
File access from web extensions is broken

### DISCUSSION THREAD

**Discussion by thegecko** [community/community#44904] [Codespaces]
### Select Topic Area

Bug

### Body

I believe a change has been made in the file scheme used in Codespaces and the github.dev versions of VS Code for web which has broken the ability for Local/Web extensions (those installed in the web extension host) to effectively use file paths or use the workspace.fs.* APIs.

For example, a debug adapter running in the web extension uses paths which are no longer recognised by the UI.

To recreate this, install the [mock-debug](https://github.com/microsoft/vscode-mock-debug) extension in the web extension host of the VSCode for web flavours and observe the differences:

**vscode.dev**
The debugger runs in the browser and accesses files using the browser filesystem API. All works as expected.

**@vscode/test-web**
This runs on a developer's machine and I believe should mimic how codespaces/github.dev work. This works as expected.

**github.dev**
The debugger fails when setting breakpoints or opening files with:

```bash
Uncaught (in promise) Error (FileSystemError): Unable to read file '/vscode-vfs://github%2b7b2276223a312c22726566223a7b2274797065223a362c226964223a226d61696e227d7d/arm-debug/mock-debug/readme.md' (Unavailable (FileSystemError): Error: No file system handle registered (/vscode-vfs:/))
```

**codespaces**
You must ensure `mock-debug` is installed "Locally" in the web extension host, (by default it is installed in the codespace host and will work correctly) by modifying the settings:

```json
{
    "remote.extensionKind": {
        "ms-vscode.mock-debug": [
            "ui"
        ]
    }
}
```
In the web extension host, similar errors are seen as `github.dev`:

```bash
Uncaught (in promise) Error (FileSystemError): Unable to read file '/workspaces/mock-debug/readme.md' (Unavailable (FileSystemError): Error: No file system handle registered (/workspaces))
```

Unfortunately codespaces and github.dev are also the only environments you cannot sideload or debug extensions in, so debugging this issue is extremely challenging.

I've tried reformatting paths prepending `vscode-remote://` and variations on the root fspath but this doesn't work.

How should a debug adapter format paths so they can be opened by the UI in codespaces/github.dev when the scheme and authority aren't present?

**Comment by thegecko** [comment]
Any comments on this issue?

**Comment by 0xNevo** [comment]
Based on the information provided, it appears that there is an issue with using file paths and workspace.fs.* APIs in Local/Web extensions when running in Codespaces or github.dev versions of VS Code for web. This may have been caused by recent changes made to the file scheme used in these environments.

To recreate this issue, you can install the mock-debug extension in the web extension host of the affected VSCode for web flavours and observe the differences in behavior.

Unfortunately, without additional details, it is difficult to provide a specific solution to this problem. However, potential solutions might include consulting the extension's documentation or seeking assistance from the extension's developer or community. Additionally, you may want to check the release notes for any recent updates to these platforms which may have caused this issue. You could also try debugging your code to identify the root cause of the problem and make any necessary modifications to accommodate the changes in these environments.

It appears that the issue is caused due to differences in file system handling between different flavors of VSCode for web. Here are some steps you can try to resolve this issue:

Make sure that the mock-debug extension is installed locally in the web extension host, by modifying the remote.extensionKind settings as shown in your example:

{
    "remote.extensionKind": {
        "ms-vscode.mock-debug": [
            "ui"
        ]
    }
}

If the above step doesn't work, try using workspace URIs instead of file paths while accessing files in your extension code. For example, instead of using /workspaces/mock-debug/readme.md, try vscode.Uri.parse('workspace:/mock-debug/readme.md').

Alternatively, if you absolutely need to use file paths, you might be able to use the workspace.fs APIs with the file scheme as a workaround. For example, instead of workspace.fs.readFile('/workspaces/mock-debug/readme.md'), try workspace.fs.readFile(vscode.Uri.file('/workspaces/mock-debug/readme.md')).

Let me know if any of these steps help resolve the issue you're facing!

When in Codespaces or github.dev, you should use "vscode.workspace.fs" API to work with files instead of the file system APIs, because in these two environments the file system is virtual and there aren't real files.

The format of the path to be used must remove the scheme prefix (for example file:///) and include the workspace URI, which is: vscode.workspace.uri.

Here's an example:

// Assuming that filePath is the path to a file
const {workspace} = vscode;
const uri = workspace.uri;
const resolvedPath = uri.path + '/' + filePath.replace(/^\/+/, "");
const document = await workspace.openTextDocument(resolvedPath);

This code gets the current workspace URI and resolves it by appending the path of the file you want to open or modify. Then, it opens the document using workspace.openTextDocument, which returns a promise that resolves to the opened document.

Note: You may need to adjust the regular expression pattern based on the expected format of the provided filePath parameter

**Comment by thegecko** [comment]
> To format paths so they can be opened by the UI in codespaces/github.dev, you can try using the vscode-uri library to convert the file path to a vscode.Uri object

Thanks, extension developers have the ability to do that when they have control over the paths (e.g. by injecting a filereader for a debugger as done here:  https://github.com/microsoft/vscode-mock-debug/blob/main/src/activateMockDebug.ts#L187).
However, this issue exists between the debug adapter and debug UI in VSCode and I can't work out how to format the correct string for the debug adapter protocol path contents as it doesn't transfer a URI object. I'm not sure it's even possible.

> If this doesn't resolve the issue, you may need to reach out to the VS Code support team or GitHub support

This is also raised on VS Code here: https://github.com/microsoft/vscode/issues/171991

I was also under the impression this discussion IS GitHub support!

> ...try using workspace URIs instead of file paths while accessing files in your extension code

This isn't possible, paths in the debug adapter protocol are strings and get passed straight to the VS Code UI

> The format of the path to be used must remove the scheme prefix

I've successfully done this with the paths for reading files:
https://github.com/thegecko/vscode-mock-debug/blob/mock-debug-test/src/activateMockDebug.ts#L215

But can't see how to modify paths coming back from the debugger :/

> Unfortunately, without additional details, it is difficult to provide a specific solution to this problem

To that end, I've created instructions and a demo example extension to recreate the issue here:

https://github.com/thegecko/vscode-web-test

I'm very keen to get an update on this or at least to confirm if the issue is with the adapter (there may be a certain path format to use), VS Code or codespaces.

**Comment by 0xNevo** [comment]
Based on the information provided, it seems like the issue is related to the format of the path being passed between the debug adapter and VS Code UI. As you mentioned, paths in the debug adapter protocol are strings and get passed straight to the VS Code UI, so it's important to use the correct path format.

One thing you can try is to check if there is a certain path format that needs to be used for the debugger adapter. You can refer to the documentation or reach out to the support team for more information.

Another possible solution is to use workspace URIs instead of file paths while accessing files in your extension code. This way, you can ensure that the correct path format is used throughout your extension code and make it easier to work with.

If none of these solutions work, it may be helpful to provide additional details or try to recreate the issue in a demo example extension. This will allow you to isolate the problem and make it easier to track down the source of the issue.

In any case, it's always a good idea to reach out to the VS Code or GitHub support team for further assistance if needed. Good luck!

**Comment by thegecko** [comment]
I can confirm this is a string formatting issue. A fix for `mock-debug` has been implemented here:

https://github.com/microsoft/vscode-mock-debug/pull/93
